### PR TITLE
[CUDA] Fix SDPA incorrect result with manual mask on long sequences

### PR DESCRIPTION
**Pull request by HongminTan**
The mem_eff attention forward kernel uses int32_t for *_strideM fields in its Params struct.  The bias pointer is computed as

    attn_bias_ptr + query_start * bias_strideM + iter_key_start

where query_start is uint32_t.  For a square causal mask with side N, bias_strideM equals N.  When N > 65536 the product overflows uint32 (65536 * 65536 = 2^32), wrapping the pointer offset and reading wrong mask data.  is_causal=True is unaffected because it uses block-level comparison instead of the bias pointer.

Widen all *_strideM from int32_t to int64_t, matching the backward kernel fix in #155397.

Test Plan:

```
CUDA_VISIBLE_DEVICES=0 python3 -c "
import torch, torch.nn.functional as F
device = 'cuda:0'; D = 128
for seq_len in [65000, 65536, 65537, 66000, 68000]:
    torch.manual_seed(42)
    q = torch.randn(1,1,seq_len,D, dtype=torch.float32, device=device)
    k = torch.randn(1,1,seq_len,D, dtype=torch.float32, device=device)
    v = torch.randn(1,1,seq_len,D, dtype=torch.float32, device=device)
    i = torch.arange(seq_len, device=device).unsqueeze(1)
    j = torch.arange(seq_len, device=device).unsqueeze(0)
    mask = i >= j; del i, j
    out_mask = F.scaled_dot_product_attention(q,k,v, attn_mask=mask)
    out_causal = F.scaled_dot_product_attention(q,k,v, is_causal=True)
    diff = (out_mask - out_causal).abs().max().item()
    print(f'seq={seq_len:>6d}: diff={diff:.6e} {\"BUG\" if diff>0.01 else \"OK\"}')
    del q,k,v,mask,out_mask,out_causal; torch.cuda.empty_cache()
"
```
Related issue: https://github.com/pytorch/pytorch/issues/162588

cc @drisspg @liangel-02 @howardzhang-cv

### CODE REVIEW

**Review by eqy — COMMENTED** [review]
Thanks, as pointed out by codex just increasing the width of the integer math used for offset calculations can introduce perf. overhead. Have you done any before/after comparisons of performance with this change?

**Comment by HongminTan** [comment]
Yes. Built two copies differing only by the `int32_t`→`int64_t` lines in this header (same compiler/flags, incremental rebuild), forward-only, `EFFICIENT_ATTENTION` backend, CUDA-event timing, median of 20 iters after
  warmup (per-iter std < 0.5 ms).

  **Setup:** 1× NVIDIA H20 (sm_90), CUDA 12.9, `b=1, heads=8, head_dim=128`.

  **`manual_mask`** (additive causal mask, goes through `bias_strideM`):

  | dtype | seq_len | before int32 (ms) | after int64 (ms) | Δ% |
  |-------|---------|-------------------|------------------|-------|
  | fp16  | 4096    | 1.553             | 1.562            | +0.53% |
  | fp16  | 8192    | 6.029             | 6.047            | +0.29% |
  | fp16  | 16384   | 23.037            | 23.173           | +0.59% |
  | fp16  | 32768   | 90.232            | 90.430           | +0.22% |
  | fp16  | 65536   | 357.453           | 357.174          | -0.08% |
  | fp16  | 131072  | 1417.665          | 1408.339         | -0.66% |
  | bf16  | 4096    | 1.549             | 1.558            | +0.57% |
  | bf16  | 8192    | 6.011             | 6.056            | +0.75% |
  | bf16  | 16384   | 23.000            | 23.228           | +0.99% |
  | bf16  | 32768   | 89.970            | 90.481           | +0.57% |
  | bf16  | 65536   | 356.451           | 357.515          | +0.30% |
  | bf16  | 131072  | 1414.808          | 1409.613         | -0.37% |
  | fp32  | 4096    | 6.134             | 6.152            | +0.29% |
  | fp32  | 8192    | 24.455            | 24.450           | -0.02% |
  | fp32  | 16384   | 94.149            | 94.286           | +0.15% |
  | fp32  | 32768   | 369.545           | 370.143          | +0.16% |
  | fp32  | 65536   | 1478.793          | 1481.413         | +0.18% |
  | fp32  | 131072  | 5881.536          | 5894.513         | +0.22% |

  **`no_mask`** (only `q/k/v/o_strideM`):

  | dtype | seq_len | before int32 (ms) | after int64 (ms) | Δ% |
  |-------|---------|-------------------|------------------|-------|
  | fp16  | 4096    | 1.243             | 1.241            | -0.15% |
  | fp16  | 8192    | 4.794             | 4.777            | -0.36% |
  | fp16  | 16384   | 18.152            | 18.184           | +0.18% |
  | fp16  | 32768   | 70.757            | 70.801           | +0.06% |
  | fp16  | 65536   | 281.748           | 282.274          | +0.19% |
  | fp16  | 131072  | 1119.404          | 1121.054         | +0.15% |
  | bf16  | 4096    | 1.240             | 1.242            | +0.15% |
  | bf16  | 8192    | 4.784             | 4.783            | -0.02% |
  | bf16  | 16384   | 18.147            | 18.228           | +0.45% |
  | bf16  | 32768   | 70.734            | 70.826           | +0.13% |
  | bf16  | 65536   | 281.715           | 282.368          | +0.23% |
  | bf16  | 131072  | 1119.381          | 1121.650         | +0.20% |
  | fp32  | 4096    | 5.652             | 5.656            | +0.06% |
  | fp32  | 8192    | 22.330            | 22.391           | +0.27% |
  | fp32  | 16384   | 86.257            | 86.485           | +0.26% |
  | fp32  | 32768   | 345.023           | 345.756          | +0.21% |
  | fp32  | 65536   | 1378.901          | 1381.979         | +0.22% |
  | fp32  | 131072  | 5486.798          | 5499.434         | +0.23% |

  Across all 36 points: mean +0.20%, median +0.21%, range -0.66% … +0.99%. No regression beyond run-to-run noise, and no growth with seq_len.

**Comment by eqy** [comment]
I also missed this PR which is being landed, is it related? https://github.com/pytorch/pytorch/pull/187225

**Comment by HongminTan** [comment]
Thanks for the review. I also noticed #187225. It appears to address the same overflow issue and additionally fixes overflow in some other offset computation paths (such as the dropout RNG offset).

My intention here was to keep the implementation consistent with the backward-side fix in #155397 by promoting the relevant offset/stride fields to int64_t. From the performance measurements I ran, I did not observe any regression. I also checked register usage with Nsight Compute on H20:

- FP32 shows a small increase in register usage, but no spilling.
- BF16 and FP16 show no increase in register pressure.

Given these results, I believe the wider types are acceptable from a performance perspective while keeping consistency with the backward path.

If needed, I can further extend the int64_t conversion to the other related offsets as well, similar to the backward implementation.

**Comment by HongminTan** [comment]
@eqy 
I noticed that PR #187225 has already been merged. If there are no further changes needed from this PR, please feel free to close it.
Thank you for your time and review.

**Comment by drisspg** [comment]
yeah closing for; https://github.com/pytorch/pytorch/pull/187225
